File: week-04-fastapi/app/routers/comments.py
from fastapi import APIRouter, HTTPException, Query
from bson import ObjectId
from datetime import datetime
from app.database import get_db
from app.schemas import CommentCreate, CommentResponse

router = APIRouter(prefix="/comments", tags=["comments"])


@router.get("", response_model=dict)
def list_comments(
    post_id: str | None = None,
    page: int = Query(1, ge=1),
    page_size: int = Query(10, ge=1, le=100),
):
    db = get_db()
    query = {}
    if post_id:
        if not ObjectId.is_valid(post_id):
            raise HTTPException(400, "Invalid post ID")
        query["postId"] = ObjectId(post_id)

    total = db.comments.count_documents(query)
    comments = (
        db.comments.find(query)
        .sort("createdAt", -1)
        .skip((page - 1) * page_size)
        .limit(page_size)
    )
    items = [CommentResponse(**c) for c in comments]
    return {"items": items, "total": total, "page": page, "page_size": page_size}


@router.get("/{comment_id}", response_model=CommentResponse)
def get_comment(comment_id: str):
    if not ObjectId.is_valid(comment_id):
        raise HTTPException(400, "Invalid comment ID")
    db = get_db()
    comment = db.comments.find_one({"_id": ObjectId(comment_id)})
    if not comment:
        raise HTTPException(404, "Comment not found")
    return CommentResponse(**comment)


@router.post("", response_model=CommentResponse, status_code=201)
def create_comment(body: CommentCreate):
    db = get_db()
    if not ObjectId.is_valid(body.postId):
        raise HTTPException(400, "Invalid post ID")
    if not ObjectId.is_valid(body.authorId):
        raise HTTPException(400, "Invalid author ID")
    if not db.posts.find_one({"_id": ObjectId(body.postId)}):
        raise HTTPException(400, "Post not found")
    if not db.users.find_one({"_id": ObjectId(body.authorId)}):
        raise HTTPException(400, "User not found")

    doc = {
        "postId": ObjectId(body.postId),
        "authorId": ObjectId(body.authorId),
        "body": body.body,
        "createdAt": datetime.now(),
    }
    result = db.comments.insert_one(doc)

    db.posts.update_one(
        {"_id": ObjectId(body.postId)},
        {"$inc": {"commentCount": 1}}
    )

    comment = db.comments.find_one({"_id": result.inserted_id})
    return CommentResponse(**comment)


@router.delete("/{comment_id}", status_code=204)
def delete_comment(comment_id: str):
    if not ObjectId.is_valid(comment_id):
        raise HTTPException(400, "Invalid comment ID")
    db = get_db()
    comment = db.comments.find_one({"_id": ObjectId(comment_id)})
    if not comment:
        raise HTTPException(404, "Comment not found")

    db.posts.update_one(
        {"_id": comment["postId"]},
        {"$inc": {"commentCount": -1}}
    )
    db.comments.delete_one({"_id": ObjectId(comment_id)})

File: week-05-express/src/routes/comments.js
import { Router } from "express";
import { ObjectId } from "mongodb";
import { getDb } from "../db.js";

const router = Router();

router.get("/", async (req, res) => {
  const db = getDb();
  const { post_id, page = 1, page_size = 10 } = req.query;
  const pageNum = Math.max(1, parseInt(page));
  const pageSize = Math.min(100, Math.max(1, parseInt(page_size)));

  const query = {};
  if (post_id) {
    if (!ObjectId.isValid(post_id)) return res.status(400).json({ detail: "Invalid post ID" });
    query.postId = new ObjectId(post_id);
  }

  const total = await db.collection("comments").countDocuments(query);
  const comments = await db.collection("comments").find(query)
    .sort({ createdAt: -1 })
    .skip((pageNum - 1) * pageSize)
    .limit(pageSize)
    .toArray();

  const items = comments.map(c => ({
    id: c._id.toString(),
    postId: c.postId.toString(),
    authorId: c.authorId.toString(),
    body: c.body,
    createdAt: c.createdAt,
  }));

  res.json({ items, total, page: pageNum, page_size: pageSize });
});

router.get("/:id", async (req, res) => {
  if (!ObjectId.isValid(req.params.id)) return res.status(400).json({ detail: "Invalid comment ID" });
  const db = getDb();
  const comment = await db.collection("comments").findOne({ _id: new ObjectId(req.params.id) });
  if (!comment) return res.status(404).json({ detail: "Comment not found" });
  res.json({
    id: comment._id.toString(),
    postId: comment.postId.toString(),
    authorId: comment.authorId.toString(),
    body: comment.body,
    createdAt: comment.createdAt,
  });
});

router.post("/", async (req, res) => {
  const db = getDb();
  const body = req.body;

  if (!ObjectId.isValid(body.postId)) return res.status(400).json({ detail: "Invalid post ID" });
  if (!ObjectId.isValid(body.authorId)) return res.status(400).json({ detail: "Invalid author ID" });
  if (!await db.collection("posts").findOne({ _id: new ObjectId(body.postId) })) {
    return res.status(400).json({ detail: "Post not found" });
  }
  if (!await db.collection("users").findOne({ _id: new ObjectId(body.authorId) })) {
    return res.status(400).json({ detail: "User not found" });
  }

  const doc = {
    postId: new ObjectId(body.postId),
    authorId: new ObjectId(body.authorId),
    body: body.body,
    createdAt: new Date(),
  };

  const result = await db.collection("comments").insertOne(doc);

  await db.collection("posts").updateOne(
    { _id: new ObjectId(body.postId) },
    { $inc: { commentCount: 1 } }
  );

  const comment = await db.collection("comments").findOne({ _id: result.insertedId });
  res.status(201).json({
    id: comment._id.toString(),
    postId: comment.postId.toString(),
    authorId: comment.authorId.toString(),
    body: comment.body,
    createdAt: comment.createdAt,
  });
});

router.delete("/:id", async (req, res) => {
  if (!ObjectId.isValid(req.params.id)) return res.status(400).json({ detail: "Invalid comment ID" });
  const db = getDb();
  const comment = await db.collection("comments").findOne({ _id: new ObjectId(req.params.id) });
  if (!comment) return res.status(404).json({ detail: "Comment not found" });

  await db.collection("posts").updateOne(
    { _id: comment.postId },
    { $inc: { commentCount: -1 } }
  );
  await db.collection("comments").deleteOne({ _id: new ObjectId(req.params.id) });
  res.status(204).send();
});

export default router;
